op-node/rollup/finality/finalizer.go

```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
-				BlockHash:      fd.L2Block.Hash.String(),
-				BlockTimestamp: fd.L2Block.Time,
+			// If there are any L2 blocks that are not finalized, then we can't finalize the later ones.
+			// This is because we need to finalize the L2 blocks in order.
+			blockCount := int(fd.L2Block.Number - finalizedL2.Number)
+			l2Blocks := make(map[uint64]eth.L2BlockRef)
+			queryBlocks := make([]*cwclient.L2Block, blockCount)
+			for i := 0; i < blockCount; i++ {
+				blockNumber := uint64(i) + finalizedL2.Number + uint64(1)
+				l2Block, err := fi.l2Fetcher.L2BlockRefByNumber(fi.ctx, blockNumber)
+				if err != nil {
+					fi.emitter.Emit(rollup.EngineTemporaryErrorEvent{Err: fmt.Errorf("failed to check if on finalizing L2 chain, could not fetch block %d: %w", blockNumber, err)})
+					return
+				}
+				l2Blocks[blockNumber] = l2Block
+
+				queryBlocks[i] = &cwclient.L2Block{
+					BlockHeight:    l2Block.Number,
+					BlockHash:      l2Block.Hash.String(),
+					BlockTimestamp: l2Block.Time,
+				}
+				fi.log.Debug(
+					"babylon gadget query params",
+					"block_height", queryBlocks[i].BlockHeight,
+					"block_hash", queryBlocks[i].BlockHash,
+					"block_timestamp", queryBlocks[i].BlockTimestamp,
+				)
+
 			}
-			fi.log.Debug(
-				"babylon gadget query params",
-				"block_height", queryParams.BlockHeight,
-				"block_hash", queryParams.BlockHash,
-				"block_timestamp", queryParams.BlockTimestamp,
-			)
-			babylonFinalized, err := fi.babylonFinalityClient.QueryIsBlockBabylonFinalized(queryParams)
-			fi.log.Debug("babylon gadget query result", "babylon_finalized", babylonFinalized)
+			// query the last finalized block in the block range on Babylon
+			lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+			fi.log.Debug("babylon gadget query result", "babylon_finalized_block_number", lastFinalizedBlockNumber)
 
 			// If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
 			// for any other error types, emit a critical error event.
-			if err != nil && !errors.Is(err, sdk.ErrNoFpHasVotingPower) {
+			if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) {
 				fi.emitter.Emit(rollup.CriticalErrorEvent{Err: fmt.Errorf("failed to check if block %d is finalized on Babylon: %w", fd.L2Block.Number, err)})
 				return
 			}
 
-			// set finalized status
-			if babylonFinalized {
-				fi.log.Debug("set finalized status", "l2_block", fd.L2Block)
-				finalizedL2 = fd.L2Block
-				finalizedDerivedFrom = fd.L1Block
+			// set finalized block(s)
+			if lastFinalizedBlockNumber != nil {
+				for finalizedBlockNumber := *lastFinalizedBlockNumber; finalizedBlockNumber > finalizedL2.Number; finalizedBlockNumber-- {
+					finalizedL2Block := l2Blocks[*lastFinalizedBlockNumber]
+					fi.log.Debug("set finalized block", "l2_block", finalizedL2Block)
+					finalizedL2 = finalizedL2Block
+					finalizedDerivedFrom = fd.L1Block
+				}
 			}
```

Comment:
could you help me understand how does this work? 

let's say:
- `finalizedL2.Number` is block 55
- `queryBlocks` has blocks from 56 to 60
- `lastFinalizedBlockNumber` is 60

then in this for loop, we:
- first set `finalizedL2` to 60
- then set `finalizedL2` to 59
- ...
- finally set `finalizedL2` to 56

which looks wrong to me. shouldn't we just set it to 60? why do we need this for loop?